### DISCUSSION TITLE
[javascript] are conditional/contextual keywords a good idea?

### DISCUSSION THREAD

**Discussion by meraedit** [antlr/grammars-v4#3309] [Q&A] [answered]
With the current grammar definition for javascript, some keywords cannot be used as identifier names, even though the browser supports them.
For instance 
`var form = 1;`
is reported as a syntax error because 'from' is a keyword.

The es standard allows it: [Keywords and Reserved Words](https://tc39.es/ecma262/#sec-keywords-and-reserved-words) 
I noticed on the Readme page that the grammar defined in the g4 file does not exactly correspond to the ECMAScript standard,
but I think it would be nice to have those in the grammar:

`conditionalKeyword
    : NonStrictLet
    | Static
    | Implements
    | Interface
    | Package
    | Private
    | Public
    | Protected
    | Async
    | Await
    | From
    | As
    | Yield
    ;
`
Attached the g4 file 
[JavaScriptParser.zip](https://github.com/antlr/grammars-v4/files/11113752/JavaScriptParser.zip)

Is there a better way to avoid getting syntax errors due to conditional keywords?

**Comment by KvanTTT** [comment, marked as answer]
You can use semantic predicates in the following way:

```g4
getter
    : {_input.LT(1).getText().equals("get")}? Identifier propertyName
    ;
```

But it's not a very good solution: it's target-specific and it looks cumbersome.

The feature for contextual keywords was suggested some time ago: [Token value comparison operator](https://github.com/antlr/antlr4/issues/1965). Unfortunately, unlikely it will be implemented in ANTLR 4.
